Add the instrument name to ExtractIlluminaRunInfo.

nh13 opened the pull request.

tfenne commented on src/main/scala/com/fulcrumgenomics/basecalling/ExtractIlluminaRunInfo.scala:
```diff
@@ -38,10 +38,11 @@ import dagr.sopt.{arg, clp}
     |2. Flowcell Barcode: the flowcell barcode.
-    |3. Run Date: the date of the sequencing run.
-    |4. Read Structure: the description of the logical structure of cycles within the sequencing run, including which cycles
+    |3. Instrument Name: the instrument name.
```
I'm just realizing that it would be nice in the usage to have the actual column names (e.g. `instrument_name`) vs. how you have it now.

nh13 commented:
@tfenne this one became more involved as I added and deprecated a few methods in `Metric`.  I also changed the `ExtractIlluminaRunInfo` tool to emit a metric.  Unfortunately, even using a secondary `String` constructor for `ReadStructure` I can't use a `ReadStructure` in a metric.  I'll think about adding support in Dagr.

tfenne commented on src/main/scala/com/fulcrumgenomics/basecalling/ExtractIlluminaRunInfo.scala:
```diff
@@ -27,74 +27,67 @@ package com.fulcrumgenomics.basecalling
+  flowcell_barcode: String,
+  instrument_name: String,
+  run_date: Iso8601Date,
+  read_structure: String, // NB: would like this to be ReadStructure
```
Does it not work?  Will Metric not toString it and use the `apply(String)` to re-hydrate it?

tfenne commented on src/main/scala/com/fulcrumgenomics/util/Metric.scala:
```diff
@@ -103,19 +103,45 @@ object Metric {
+  /** Writes a metric to the given writer.  The first line will be a header with the field names.  Each subsequent
+    * line is a single metric.
+    */
+  def write[T <: Metric](writer: Writer, metrics: TraversableOnce[T])(implicit tt: ru.TypeTag[T]): Unit = write(writer, metrics.toSeq:_*)
```
Something doesn't feel awesome about accepting a `TraversableOnce` and then immediately calling `toSeq` on it.  I.e. if I provided an `Iterator[MyMetric]` that produced a ton of metrics, I might get surprised by this.

nh13 replied:
Good point.  I'll fix this but note that I am not fixing up metric reading, which also uses `Seq` and doesn't have a `MetricReader`.

nh13 commented:
@tfenne one last look?

tfenne commented on src/main/scala/com/fulcrumgenomics/basecalling/ExtractIlluminaRunInfo.scala:
```diff
@@ -27,74 +27,67 @@ package com.fulcrumgenomics.basecalling
+  flowcell_barcode: String,
+  instrument_name: String,
+  run_date: Iso8601Date,
+  read_structure: ReadStructure, // NB: would like this to be ReadStructure
```
Comment can go now.

tfenne commented on src/main/scala/com/fulcrumgenomics/util/ReadStructure.scala:
```diff
@@ -31,13 +31,16 @@ import scala.reflect.ClassTag
+  def apply(readStructure: String): ReadStructure = new ReadStructure(segs=segments(readStructure=readStructure))
+
+  /** Creates a sequence of read segments from a string. */
+  private[ReadStructure] def segments(readStructure: String): Seq[ReadSegment] = {
```
Isn't `private[ReadStructure]` the same thing as `private` here?

tfenne commented on src/main/scala/com/fulcrumgenomics/util/ReadStructure.scala:
```diff
@@ -94,6 +97,9 @@ object ReadStructure {
 class ReadStructure(segs: Seq[ReadSegment], resetOffsets: Boolean = false) extends immutable.Seq[ReadSegment] {
   import ReadStructure.SubRead
+
+  def this(readStructure: String) = this(ReadStructure.segments(readStructure=readStructure))
```
Can you just add a one-line comment here please?
